File: src/main/java/com/anod/appwatcher/AppWatcherApplication.java
package com.anod.appwatcher;

import android.app.Application;
import android.content.Context;
import android.content.res.Configuration;
import android.support.v7.app.AppCompatDelegate;
import android.view.ViewConfiguration;

import com.android.volley.VolleyLog;
import com.anod.appwatcher.utils.AppDetailsUploadDate;
import com.anod.appwatcher.utils.MetricsManagerEvent;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.crash.FirebaseCrash;
import com.squareup.leakcanary.LeakCanary;

import java.lang.reflect.Field;
import java.util.Locale;

import info.anodsplace.android.log.AppLog;

public class AppWatcherApplication extends Application implements AppLog.Listener {
    private ObjectGraph mObjectGraph;

    @Override
    public void onCreate() {
        super.onCreate();
        LeakCanary.install(this);

        try {
            ViewConfiguration config = ViewConfiguration.get(this);
            Field menuKeyField = ViewConfiguration.class.getDeclaredField("sHasPermanentMenuKey");
            if (menuKeyField != null) {
                menuKeyField.setAccessible(true);
                menuKeyField.setBoolean(config, false);
            }
        } catch (Exception ex) {
            // Ignore
        }

        AppLog.LOGGER = new FirebaseLogger();
        AppLog.setDebug(true, "AppWatcher");
        AppLog.instance().setListener(this);
        //VolleyLog.setTag("AppWatcher");

        mObjectGraph = new ObjectGraph(this);
        Preferences prefs = new Preferences(this);
        AppCompatDelegate.setDefaultNightMode(prefs.getNightMode());
    }

    public ObjectGraph getObjectGraph() {
        return mObjectGraph;
    }

    public static AppWatcherApplication get(Context context) {
        return (AppWatcherApplication) context.getApplicationContext();
    }

    public static ObjectGraph provide(Context context) {
        return ((AppWatcherApplication) context.getApplicationContext()).getObjectGraph();
    }

    public boolean isNightTheme() {
        int currentNightMode = getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;
        switch (currentNightMode) {
            case Configuration.UI_MODE_NIGHT_YES:
                return true;
            case Configuration.UI_MODE_NIGHT_NO:
            case Configuration.UI_MODE_NIGHT_UNDEFINED:
                return false;
        }
        return false;
    }

    @Override
    public void onLogException(Throwable tr) {
        FirebaseCrash.report(tr);
        if (tr instanceof AppDetailsUploadDate.ExtractDateError) {
            AppDetailsUploadDate.ExtractDateError error = (AppDetailsUploadDate.ExtractDateError) tr;
            MetricsManagerEvent.track(this, "error_extract_date",
                    "LOCALE", error.locale,
                    "DEFAULT_LOCALE", error.defaultlocale,
                    "ACTUAL", error.actual,
                    "EXPECTED", error.expected,
                    "EXPECTED_FORMAT", error.expectedFormat,
                    "CUSTOM", error.isCustomParser? "YES" : "NO");
        }
    }

    private class FirebaseLogger extends AppLog.Logger.Android {
        @Override
        public void println(int priority, String tag, String msg) {
            super.println(priority, tag, msg);
            FirebaseCrash.logcat(priority, tag, msg);
        }
    }
}

File: app/src/main/java/com/anod/appwatcher/AppWatcherApplication.kt
package com.anod.appwatcher

import android.app.Application
import android.content.Context
import android.os.Build
import android.os.StrictMode
import androidx.appcompat.app.AppCompatDelegate
import androidx.room.Room
import androidx.work.Configuration
import com.anod.appwatcher.backup.createBackupModule
import com.anod.appwatcher.backup.gdrive.UploadServiceContentObserver
import com.anod.appwatcher.database.AppsDatabase
import com.anod.appwatcher.installed.createInstalledModule
import com.anod.appwatcher.preferences.Preferences
import com.anod.appwatcher.sync.SyncNotification
import com.anod.appwatcher.utils.networkConnection
import com.google.firebase.crashlytics.FirebaseCrashlytics
import finsky.api.DfeError
import info.anodsplace.applog.AndroidLogger
import info.anodsplace.applog.AppLog
import info.anodsplace.context.ApplicationInstance
import java.io.IOException
import java.net.UnknownHostException
import org.koin.core.component.KoinComponent
import org.koin.core.component.get
import org.koin.core.context.startKoin
import org.koin.dsl.module

class AppWatcherApplication : Application(), AppLog.Listener, ApplicationInstance, Configuration.Provider, KoinComponent {

    override val workManagerConfiguration = Configuration.Builder().apply {
        setMinimumLoggingLevel(android.util.Log.DEBUG)
    }.build()

    override val appCompatNightMode: Int
        get() = get<Preferences>().appCompatNightMode

    val appsDatabase: AppsDatabase by lazy {
        Room.databaseBuilder(this, AppsDatabase::class.java, AppsDatabase.dbName)
            .enableMultiInstanceInvalidation()
            .addMigrations(*AppsDatabase.migrations)
            .build()
    }

    override fun onCreate() {
        super.onCreate()

        if (BuildConfig.DEBUG) {
            StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder()
                .detectActivityLeaks()
                .detectAll()
                .detectNonSdkApiUsage()
                .penaltyLog()
                .build())
        }

        AppLog.setDebug(BuildConfig.DEBUG, "AppWatcher")

        startKoin {
            modules(modules = listOf(
                module {
                    single<Context> { this@AppWatcherApplication }
                    single<Application> { this@AppWatcherApplication }
                    single { appsDatabase }
                },
                createAppModule(),
                createPlayStoreModule(),
                createBackupModule(),
                createInstalledModule()
            ))
        }

        val prefs = get<Preferences>()
        if (prefs.collectCrashReports) {
            FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(true)
            AppLog.logger = FirebaseLogger()
            AppLog.instance.listener = this
        }

        if (prefs.isDriveSyncEnabled) {
            appsDatabase.invalidationTracker.addObserver(get<UploadServiceContentObserver>())
        }

        AppCompatDelegate.setDefaultNightMode(prefs.appCompatNightMode)
        SyncNotification(info.anodsplace.context.ApplicationContext(this), get()).createChannels()
    }

    override fun onLogException(tr: Throwable) {
        if (isNetworkError(tr) || tr is kotlinx.coroutines.CancellationException) {
            // Ignore
            return
        }

        if (get<Preferences>().collectCrashReports) {
            FirebaseCrashlytics.getInstance().recordException(tr)
        }
    }

    private fun isNetworkError(tr: Throwable): Boolean {
        return tr is DfeError
            || tr is UnknownHostException
            || (tr is IOException && tr.message?.contains("NetworkError") == true)
            || networkConnection.isNetworkException(tr)
    }

    private inner class FirebaseLogger : AndroidLogger() {
        override fun println(priority: Int, tag: String, msg: String) {
            super.println(priority, tag, msg)
            FirebaseCrashlytics.getInstance().log("$priority/$tag: $msg")
        }
    }
}